stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -1007,3 +1007,77 @@ function matmul3x3!(C::AbstractMatrix, tA, tB, A::AbstractMatrix, B::AbstractMat
     end # inbounds
     C
 end
+
+# Three-argument *
+"""
+    *(A, B::AbstractMatrix, C)
+
+Most 3-argument `*` calls containing matrices or vectors are done in an efficient way,
+rather than the left-to-right default of `*(x,y,z,...)`.
+
+This can mean performing `B*C` first if `C::AbstractVector`, calling `dot`, using
+five-argument `mul!` to fuse the scalar `A::Number` with the matrix multiplication `B*C`,
+or examining `size.((A,B,C))` to choose which to multiply first.
+
+!!! compat "Julia 1.6"
+    These optimisations require least Julia 1.6.
+"""
+*(A::AbstractMatrix, B::AbstractMatrix, x::AbstractVector) = _mat_mat_vec(A,B,x)
+
+*(A::AbstractMatrix, x::AbstractVector, γ::Number) = mat_vec_scalar(A,x,γ)
+*(A::AbstractMatrix, B::AbstractMatrix, γ::Number) = mat_mat_scalar(A,B,γ)
+*(α::Union{Real,Complex}, B::AbstractMatrix, C::AbstractVector) = mat_vec_scalar(B,C,α)
+*(α::Union{Real,Complex}, B::AbstractMatrix, C::AbstractMatrix) = mat_mat_scalar(B,C,α)
+
+*(α::Number, u::AbstractVector, tv::AdjOrTransAbsVec) = broadcast(*, α, u, tv)
+*(u::AbstractVector, tv::AdjOrTransAbsVec, γ::Number) = broadcast(*, u, tv, γ)
+*(u::AbstractVector, tv::AdjOrTransAbsVec, C::AbstractMatrix) = u * (tv*C)
+
+*(A::AbstractMatrix, B::AbstractMatrix, C::AbstractMatrix) = _tri_matmul(A,B,C)
+*(tv::AdjOrTransAbsVec, B::AbstractMatrix, C::AbstractMatrix) = (tv*B) * C
+
+_mat_mat_vec(A::AdjointAbsVec{<:Number}, B, x) = dot(A.parent, B, x)
+_mat_mat_vec(A::TransposeAbsVec{<:Real}, B, x) = dot(A.parent, B, x)
+_mat_mat_vec(A, B, x) = A * (B*x)
+
+function _tri_matmul(A,B,C)
+    n,m = size(A)
+    # m,k == size(B)
+    k,l = size(C)
+    costAB_C = n*m*k + n*k*l
+    costA_BC = m*k*l + n*m*l
+    if costA_BC < costAB_C
+        A * (B*C)
+    else
+        (A*B) * C
+    end
+end
+
+# The fast path for two arrays * one scalar is opt-in.
+_SafeMatrix{T} = Union{StridedMatrix{T}, AdjOrTransAbsMat{T, <:StridedMatrix}}
```

Comment:
Is there an existing type that could be used here, or a better description of what's needed?

Ideally this would catch every matrix whose `*` goes to `mul!(similar(A), ...)`, as those ones are safe to send to `_mat_vec_scalar(A, x, γ)`. But must exclude (for instance) StaticArrays (no mutation) and NamedDimsArrays (wrapper not handled by `similar`).